### PR TITLE
remove dbus prefix

### LINKED ISSUE #477: Remove `dbus_` prefix from macro names
There is no need for this prefix and there is no collision possible in `zbus_macros` nor `zbus`. 

We should definitely keep the existing names but deprecate them.

### LINKED ISSUE #478: macros should use `zbus` attributes rather than `dbus_proxy` and `dbus_interface`
There is no need for the long names.

We should continue to support `dbus_proxy` and `dbus_interface` aliases too, though.

### PR DESCRIPTION
**Pull request by TTWNO**
This removes the `dbus_` prefix before macros and macro attributes.

Fixes #477 

Fixes #478

### CODE REVIEW

**Comment by TTWNO** [comment]
This should fix #477 and #478

**Comment by TTWNO** [comment]
Oh, is it possible I misread it initially?

You want the attributes to be prefixed with `zbus_` and the macros to be un-prefixed, yes?

**Comment by zeenix** [comment]
> Oh, is it possible I misread it initially?
> 
> You want the attributes to be prefixed with `zbus_` 

No, just `zbus`, similar to serde's attributes (I think zvariant too).

> and the macros to be un-prefixed, yes?

Correct.

**Comment by zeenix** [comment]
> This should fix #477 and #478
> 

Nice! Please add `Fixes..` line to the description of the PR so the issues get automatically closed when it's merged.

Thanks for doing this work. I'll try to review soon..

**Comment by zeenix** [comment]
Also kindly squash fixups to the relevant commits and force push, as per the contribution guide.

**Comment by TTWNO** [comment]
I will sqash in the AM (MDT)

**Comment by zeenix** [comment]
> I will sqash in the AM (MDT)

I know life can get in the way so no pressure but just wanted to ping you, in case your forgot. :)

**Comment by zeenix** [comment]
btw, there's also [a formatting issue](https://github.com/dbus2/zbus/actions/runs/6592748039/job/17914277019?pr=496) breaking the CI.

**Comment by TTWNO** [comment]
Ddn't forget, jus time getting away from me.

**Comment by TTWNO** [comment]
I notice you mention in some of the `TODO`s that you want it to look like `serde`. What exactly did you mean by that?

**Comment by TTWNO** [comment]
Ah, I missed one thing, let me fix that up.

**Comment by TTWNO** [comment]
Ok, so I know you want me to keep the `dbus_*` crate names for the attributes, so that's the next thing I'm working on, but all attributes can be written as `#[zbus(...)]` now.

**Comment by TTWNO** [comment]
Do you want both so it's not a breaking change? So attributes could be either `zbus(...)` or `dbus_*(...)`

**Comment by zeenix** [comment]
> Do you want both so it's not a breaking change? So attributes could be either `zbus(...)` or `dbus_*(...)`

Yes. I want these changes to be non-breaking. We should also emit a warning for use of existing names that they're deprecated. See commit 5db2d13fa80fd68f8f7c9178259cd5a18a3e00ea where we did something similar using `eprintln`.

**Comment by TTWNO** [comment]
Alright, all done for `dbus_poxy/proxy` and `dbus_interface/interface`.

Is using `eprintln!` preferable to using `#[deprecated = "..."]`? Or does it not matter?

**Comment by zeenix** [comment]
> Alright, all done for `dbus_poxy/proxy` and `dbus_interface/interface`.

Nice! Thanks.

> Is using `eprintln!` preferable to using `#[deprecated = "..."]`? Or does it not matter?

Actually `deprecated` is even better, as long as it actually works. I think the reason we used `eprintln!` in 5db2d13fa80fd68f8f7c9178259cd5a18a3e00ea was that it was about attributes inside the macro.

**Comment by TTWNO** [comment]
> Actually deprecated is even better, as long as it actually works. I think the reason we used eprintln! in https://github.com/dbus2/zbus/commit/5db2d13fa80fd68f8f7c9178259cd5a18a3e00ea was that it was about attributes inside the macro.

Yes, that's right. I fear that may be the only solution for `dbus_error`, which is still not done yet due to it being slightly more complex. I'll send a message when I think it's ready for CI, no need to keep running it for now :)

**Comment by TTWNO** [comment]
Ok, it looks like a clippy warning might be coming from `zvariant`, but it looks like some exports that are meant to be there.

```
warning: unused import: `from_value::*`
  --> zvariant/src/lib.rs:85:9
   |
85 | pub use from_value::*;
   |         ^^^^^^^^^^^^^
   |
   = note: `#[warn(unused_imports)]` on by default

warning: unused import: `into_value::*`
  --> zvariant/src/lib.rs:88:9
   |
88 | pub use into_value::*;
   |         ^^^^^^^^^^^^^
```

**Comment by TTWNO** [comment]
Of course, it's easy to just remove the exports... but it looks purposeful.

**Comment by zeenix** [comment]
> > Actually deprecated is even better, as long as it actually works. I think the reason we used eprintln! in [5db2d13](https://github.com/dbus2/zbus/commit/5db2d13fa80fd68f8f7c9178259cd5a18a3e00ea) was that it was about attributes inside the macro.
> 
> Yes, that's right. I fear that may be the only solution for `dbus_error`,

Oh, nobody said anything about `DBusError`. I'm not sure what can be done about that because `zbus::Error` is already taken and their both different things. So I decided to leave it be.

> Ok, it looks like a clippy warning might be coming from `zvariant`

I think I fixed these recently. Can you try rebasing on current main and see if it helps?

**Comment by zeenix** [comment]
Oh, I see you squashed both commits now. :( This should be at least 2 commits, with their own justification and `Fixes #ISSUE` in their commit log. Which is why there were two issues created.

**Comment by zeenix** [comment]
> > Ok, it looks like a clippy warning might be coming from `zvariant`
> 
> I think I fixed these recently. Can you try rebasing on current main and see if it helps?

Seems you did that already? The CI passes now. :partying_face: Now, if only you hadn't squashed the 2 commits and instead addressed my comments, I could have merged already. Could you please split the commits and redo the logs for each?

While at it, you might also want to add emoji prefixes as the [ contribution guide](https://github.com/dbus2/zbus/blob/main/CONTRIBUTING.md#submitting-pull-requests). :)

**Comment by TTWNO** [comment]
Ahh, bugger! Just realized I'm still not done.

Only the `dbus_error` derive attribute and variant attribute have warnings. `dbus_interface` and `dbus_method` were renamed, with warnings for the old proc macros, but the old attribute names could not be used. It would just error.

Renaming is easy, supporting an old way *and* a new way, not so

**Comment by zeenix** [comment]
> Renaming is easy, supporting an old way _and_ a new way, not so

I imagined it so. Anything involving the macros is hard. I myself have been editing the macros recently for #395 and it's been difficult to say the least. Anyway, I'm very thankful for your help on this and willingness to take this to completion. :+1:

**Comment by zeenix** [comment]
@TTWNO any update on this?

**Comment by TTWNO** [comment]
Ok, I think I got it all to work. I realize this is in one commit instead of two. That will have to be fixed another day.

It also got quite a bit more complex. In order not to re-implement a bunch of logic, we have to match on a generic `T` and check if it is the old or new `MethodAttributes` for example, then emit the warning when the `T::Old` variant is matched on.

It ain't pretty, but should be pretty easy to remove once the time comes.

* [X] The `dbus_` prefix has been removed from the generative proc-macros.
* [X] The method and variant macros have been re-written to all use the `zbus` name, followed by various properties within it, as before.

Let me know if there's anything else to do other than split the changes into two commits. I have example code here: https://github.com/TTWNO/zbus_testers (it assumes it is cloned as a sibling of a local copy of `zbus`, on this branch).

Phew! What a monstor!

**Review by zeenix — CHANGES_REQUESTED** [review]
Great work in general! Thanks for sticking with it. Very few people do that.

Apart from the mostly nit picks, yes, this should be **at least** 2 commits and the commit logs need to be re-written. :)

**Comment by zeenix** [comment]
Seems [some use](https://github.com/dbus2/zbus/actions/runs/7040963899/job/19175804316?pr=496) needs to be update.

**Comment by TTWNO** [comment]
Oh wow! I totally missed the feedback here.

I think I saw somewhere that this issue is being untagged from the 4.0 release. I'm totally fine with that and agree with the fact that it does not need to be waited on since it shouldn't break anything.

File: zbus_macros/src/iface.rs
Comment (TTWNO) [reply zbus_macros/src/iface.rs:17]: Can you explain what you mean here? I'm not sure I understand what you mean.

All mods are consistent in this PR, including the `mod old { ... }` available within each module's top level.

Perhaps I'm missing something.

Comment (zeenix) [reply zbus_macros/src/iface.rs:17]: > All mods are consistent in this PR, including the `mod old { ... }` available within each module's top level.

The one in the `error.rs` is called `dbus_old_error`, here it's `old`.

Comment (zeenix) [reply zbus_macros/src/iface.rs:17]: Sorry, I think the confusion was because of my typo. I hope it's clear now.

**Comment by zeenix** [comment]
For resolving conflict with `main` branch, please rebase on `main`, not merge it.

**Comment by TTWNO** [comment]
Rebased to not conflict with main.

Tests pass, and my example crate with both and new notation work as well.

Clippy and formatting pass as well.

If I'm reading our history right, the only thing left is to get this into at least two commits. I know I'm asking for a lot here, but is there any chance we could do the majority of the changes as one commit?

- `🐛 [zm] Rename attribute and derive macros; fix #477, fix #478`, preceded by
- `✨ [zd] Add AttrParse trait, old_new! macro for compatibility with deprecated macro names` 

If that doesn't work for you, then it still may take more time for me to untangle the overlapping changed.

**Comment by zeenix** [comment]
@TTWNO CI is broken btw. The good news is that, now with your other PR merged, you're not longer a first time contributor and shouldn't need approval for CI pipeline run. :)

**Comment by zeenix** [comment]
> If the split turns out be harder than I imagine this to be, just let me know and I'll consider pushing as one commit.

How did it go? I can also help by doing the git surgery for you, if you ran into issues with my recommended steps.

**Comment by TTWNO** [comment]
> How did it go? I can also help by doing the git surgery for you, if you ran into issues with my recommended steps.

Going to give it a go today.... once I rebase again :)

**Comment by TTWNO** [comment]
Ah! rebased from *my* main, not yours.

**Comment by TTWNO** [comment]
Going for gold here, just got tests all passing. Now for the git surgery.

**Comment by TTWNO** [comment]
Going to be honest, this looks like a nightmare to de-tangle. If you are able to do the surgery, I'd appreciate it. I've narrowed it down to one commit.

**Comment by TTWNO** [comment]
Ah, damn! It took the messages I was putting in the description instead of the ones I put in the `pick/squash` block. I think I'm confused about how this all works...

**Comment by zeenix** [comment]
> just got tests all passing.

Awesome. :+1: 

> Going to be honest, this looks like a nightmare to de-tangle

I can see how it can be very daunting for people not too familiar with git workflows but that's why I tried my best to give you very specific instructions to follow.

> Ah, damn! It took the messages I was putting in the description instead of the ones I put in the `pick/squash` block. I think I'm confused about how this all works...

No idea what that means but if you follow [my instructions](https://github.com/dbus2/zbus/pull/496#issuecomment-1872935885), it will be like writing a completely new commit message for each commit. No rebasing required even.

Anyway, I can take it over from here. I'll force push directly to his PR's branch so please don't touch it w/o letting me know (here or on [our matrix channel](https://matrix.to/#/#zbus:matrix.org)).

File: zbus/src/fdo.rs
Comment (zeenix) [inline zbus/src/fdo.rs:828]: what happened here? :open_mouth: The logical choice for `dbus_error` renaming would be `error` for the macro itself and `dbus_error(zbus_error)` should become `zbus(zbus_error)` (actually come to think of it, perhaps `zbus(fallthrough)` might be even better). I'll fix this as part of my rebase. The rebasing will taking much longer than expected beause of this and some other unrelated and even some unwanted changes I am seeing.

Comment (zeenix) [reply zbus/src/fdo.rs:828]: ah, nm the point about the macro name. I just keep forgetting the macro is `DBusError` since it's a derive macro unlike proxy and interface.

File: zbus_macros/src/proxy.rs
Comment (zeenix) [inline zbus_macros/src/proxy.rs:273]: why initialize all these just to ignore? :thinking:

Comment (zeenix) [reply zbus_macros/src/proxy.rs:273]: This is a big advantage of keeping commits atomic: it's easy to see if unwanted changes got in.

File: zbus_macros/iface.txt
Comment (zeenix) [inline zbus_macros/iface.txt:1]: I'm assuming this whole file was some kind of mistake? I'll remove it in my rebase.

**Comment by zeenix** [comment]
Seems I'll need a few things to fix still and I wanted to also ensure that each commit is tested independent and that deprecated names still work as expected. I'll push now what I have already and fix the rest tomorrow (and during the weekend, if I don't manage to finish it already tomorrow). Stay tuned!

**Comment by zeenix** [comment]
@TTWNO Not only `dbus_proxy` doesn't work anymore, if you use it, you get warnings about the attribute as well even if you only use `zbus` attribute everywhere. With this change:

```diff
diff --git a/zbus/tests/e2e.rs b/zbus/tests/e2e.rs
index 730d576b..ddb81ff5 100644
--- a/zbus/tests/e2e.rs
+++ b/zbus/tests/e2e.rs
@@ -22,10 +22,9 @@ use zbus::{
 use zvariant::{DeserializeDict, Optional, OwnedValue, SerializeDict, Str, Type, Value};
 
 use zbus::{
-    connection, interface,
+    connection, dbus_proxy, interface,
     message::Header,
     object_server::{InterfaceRef, SignalContext},
-    proxy,
     proxy::CacheProperties,
     Connection, ObjectServer,
 };
@@ -52,7 +51,7 @@ pub struct RefType<'a> {
     field1: Str<'a>,
 }
 
-#[proxy(assume_defaults = true, gen_blocking = true)]
+#[dbus_proxy(assume_defaults = true, gen_blocking = true)]
 trait MyIface {
     fn ping(&self) -> zbus::Result<u32>;
 

```

I get this:

```rust
The `#[dbus_proxy(...)]` attribute of proxy has been deprecated in favor of `#[zbus(...)]`.
The `#[dbus_proxy(...)]` attribute of proxy has been deprecated in favor of `#[zbus(...)]`.
// Repeats 149 times!
error: use of deprecated macro `dbus_proxy`: Use #[proxy(...)] instead.
  --> zbus/tests/e2e.rs:54:3
   |
54 | #[dbus_proxy(assume_defaults = true, gen_blocking = true)]
   |   ^^^^^^^^^^
   |
   = note: `-D deprecated` implied by `-D warnings`
   = help: to override `-D warnings` add `#[allow(deprecated)]`

error: implementation of `Deserialize` is not general enough
  --> zbus/tests/e2e.rs:54:1
   |
54 | #[dbus_proxy(assume_defaults = true, gen_blocking = true)]
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ implementation of `Deserialize` is not general enough
   |
   = note: `RefType<'_>` must implement `Deserialize<'0>`, for any lifetime `'0`...
   = note: ...but `RefType<'_>` actually implements `Deserialize<'1>`, for some specific lifetime `'1`

error[E0599]: no method named `cached_count` found for struct `MyIfaceProxy` in the current scope
   --> zbus/tests/e2e.rs:653:22
    |
54  | #[dbus_proxy(assume_defaults = true, gen_blocking = true)]
    | ---------------------------------------------------------- method `cached_count` not found for this struct
...
653 |     assert_eq!(proxy.cached_count()?, None);
    |                      ^^^^^^^^^^^^ help: there is a method with a similar name: `set_count`

error[E0599]: no method named `receive_count_changed` found for struct `MyIfaceProxy` in the current scope
   --> zbus/tests/e2e.rs:772:18
    |
54  | #[dbus_proxy(assume_defaults = true, gen_blocking = true)]
    | ---------------------------------------------------------- method `receive_count_changed` not found for this struct
...
772 |     my_obj_proxy.receive_count_changed().await;
    |                  ^^^^^^^^^^^^^^^^^^^^^ method not found in `MyIfaceProxy<'_>`

error[E0599]: no method named `cached_count` found for struct `MyIfaceProxy` in the current scope
   --> zbus/tests/e2e.rs:775:18
    |
54  | #[dbus_proxy(assume_defaults = true, gen_blocking = true)]
    | ---------------------------------------------------------- method `cached_count` not found for this struct
...
775 |     my_obj_proxy.cached_count()?;
    |                  ^^^^^^^^^^^^ help: there is a method with a similar name: `set_count`

error[E0599]: no method named `cached_count` found for struct `MyIfaceProxy` in the current scope
   --> zbus/tests/e2e.rs:777:29
    |
54  | #[dbus_proxy(assume_defaults = true, gen_blocking = true)]
    | ---------------------------------------------------------- method `cached_count` not found for this struct
...
777 |     assert_eq!(my_obj_proxy.cached_count()?, Some(0));
    |                             ^^^^^^^^^^^^ help: there is a method with a similar name: `set_count`
```

**Comment by zeenix** [comment]
BTW, I moved both issue back on the agenda for 4.0, which I really want to rollout early-mid next week (I took the week off from work for that) as I realized it won't be nice to deprecate the most commonly used API in a stable release and we're close here, despite the issues.

I think something that would help is:

* supporting existing attributes only with the existing (deprecated) macros.
* mention the new attribute in the macro deprecation warning.
* because of the above 2, no need for deprecation warning for attributes.

@TTWNO Can you handle that before next week? If not, let me know and I'll see if I can do the rest. I'll not touch your branch anymore until you let me know.

**Comment by zeenix** [comment]
@TTWNO oh and feel free to add commits on top. I'll squash them into the appropriate commit for you. Please just try your best to keep them as small as possible, so it's easy to squash.

**Comment by TTWNO** [comment]
> Can you handle that before next week?

Yes. It may come in at the last minute. But this will be done before Sunday evening (MST).

Super happy to see 4.0 coming together :)

**Comment by zeenix** [comment]
> > Can you handle that before next week?
> 
> Yes. It may come in at the last minute. But this will be done before Sunday evening (MST).

Awesome. I hope you don't hate me for this but thinking more about the new plan, I think it's now just one commit since we're no longer renaming the attribute for existing macros. So don't worry about splitting commits anymore. I wish I had come up with this plan before but I only got the idea after I tried the changes.

**Comment by TTWNO** [comment]
On this right now :)

**Comment by TTWNO** [comment]
> supporting existing attributes only with the existing deprecated macros.
> mention the new attribute in the macro deprecation warning.
> because of the above 2, no need for deprecation warning for attributes.

Done for interface and proxy macros. That error macro has me stumped.

**Comment by zeenix** [comment]
> Done for interface and proxy macros.

Great, thanks. The CI is still failing though. Also, I'm sure you can squash commits into 1 (you did that recently). :wink: 

> That error macro has me stumped.

You mean what I commented or how to handle it? :thinking:

**Comment by TTWNO** [comment]
> You mean what I commented or how to handle it? 🤔

How to handle it.

**Comment by zeenix** [comment]
> > You mean what I commented or how to handle it? 🤔
> 
> How to handle it.

If you could elaborate a little, I might be able to help.

**Comment by zeenix** [comment]
> > > You mean what I commented or how to handle it? 🤔
> > 
> > 
> > How to handle it.
> 
> If you could elaborate a little, I might be able to help.

I think I know what you mean. Given that we're not renaming `DBusError` and that it's not used as often as the other macros, I think we could just break the API here and rename the attributes w/o providing backwards compat.

**Comment by zeenix** [comment]
> I think I know what you mean. Given that we're not renaming `DBusError` and that it's not used as often as the other macros, I think we could just break the API here and rename the attributes w/o providing backwards compat.

And this one can and should go in a separate commit.

**Comment by TTWNO** [comment]
Got it. I'll get that dome this morning.

The rename should be to `Error`, yes?

**Comment by zeenix** [comment]
> Got it. I'll get that dome this morning.

:+1: 

> The rename for the derive should be to `Error`, yes?

Just to be sure, I don't end up causing any misunderstanding:

* the derive macro itself does **not** change name as it's identical to the trait it implements.
* the attribute of the macro get renamed from `dbus_error` to `zbus`.
* the `zbus_error` value of `zbus` attribute should just become `error`, i-e `#[zbus(error)]`.

**Comment by TTWNO** [comment]
Thanks for being patient. Answering you in class to make sure this gets into 4.0 :)

**Comment by TTWNO** [comment]
This is now a breaking change, due to `dbus_error(zbus)` has been *replaced* with `zbus(error)`; no deprecations.

**Comment by zeenix** [comment]
> Thanks for being patient. 

Thanks for your patience too. I'm guessing I should handle the git surgery? :)

> Answering you in class to make sure this gets into 4.0 :)

:+1:

**Comment by zeenix** [comment]
@TTWNO btw, I would invite you to [our matrix channel](https://matrix.to/#/#zbus:matrix.org) as there are communications that are best handled on chat, rather than here.

**Review by zeenix — APPROVED** [review]
Now that I've modified your changes and split into commits, I approve it. :)

**Pull request merged**
